storage/client/src/main/java/com/azure/storage/queue/QueueAsyncClientBuilder.java

```diff
@@ -1,5 +1,194 @@
+
+/**
+ * Fluent builder for queue async clients.
+ */
+public final class QueueAsyncClientBuilder {
+    private final List<HttpPipelinePolicy> policies;
+
+    private URL endpoint;
+    private String queueName;
+    private SASTokenCredential sasTokenCredential;
+    private SharedKeyCredential sharedKeyCredential;
+    private HttpClient httpClient;
+    private HttpLogDetailLevel logLevel;
+    private RetryPolicy retryPolicy;
+    private Configuration configuration;
+
+    QueueAsyncClientBuilder() {
+        retryPolicy = new RetryPolicy();
+        logLevel = HttpLogDetailLevel.NONE;
+        policies = new ArrayList<>();
+
+        configuration = ConfigurationManager.getConfiguration();
+    }
+
+    /**
+     * Constructs an instance of QueueAsyncClient based on the configurations stored in the builder.
+     * @return a new client instance
+     */
+    public QueueAsyncClient build() {
+        Objects.requireNonNull(endpoint);
+        Objects.requireNonNull(queueName);
+
+        if (sasTokenCredential == null && sharedKeyCredential == null) {
+            throw new IllegalArgumentException("Credentials are required for authorization");
+        }
+
+        // Closest to API goes first, closest to wire goes last.
+        final List<HttpPipelinePolicy> policies = new ArrayList<>();
+
+        policies.add(new UserAgentPolicy(QueueConfiguration.NAME, QueueConfiguration.VERSION, configuration));
+        policies.add(new RequestIdPolicy());
+        policies.add(new AddDatePolicy());
+
+        if (sharedKeyCredential != null) {
+            policies.add(new SharedKeyCredentialPolicy(sharedKeyCredential));
+        } else {
+            policies.add(new SASTokenCredentialPolicy(sasTokenCredential));
+        }
+
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(retryPolicy);
+
+        policies.addAll(this.policies);
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+        policies.add(new HttpLoggingPolicy(logLevel));
+
+        HttpPipeline pipeline = HttpPipeline.builder()
+            .policies(policies.toArray(new HttpPipelinePolicy[0]))
+            .httpClient(httpClient)
+            .build();
+
+        return new QueueAsyncClient(endpoint, pipeline, queueName);
+    }
+
+    /**
+     * Sets the service endpoint, additionally parses it for information (SAS token, queue name)
+     * @param endpoint URL of the service
+     * @return the updated QueueAsyncClientBuilder object
+     * @throws IllegalArgumentException If {@code endpoint} isn't a proper URL
+     */
+    public QueueAsyncClientBuilder endpoint(String endpoint) {
+        Objects.requireNonNull(endpoint);
+        try {
+            URL fullURL = new URL(endpoint);
+            this.endpoint = new URL(fullURL.getProtocol() + "://" + fullURL.getHost());
+
+            // Attempt to get the queue name from the URL passed
+            String[] pathSegments = fullURL.getPath().split("/", 2);
```

Comment:
Investigate if https://github.com/Azure/azure-sdk-for-java/blob/master/core/azure-core/src/main/java/com/azure/core/implementation/http/UrlBuilder.java can be used to simplify stuff here.

Reply:
Doesn't look that it would simplify it much given the path in UrlBuilder is stored as a string and not in a linked list or another structure where the path pieces are split.